apps/base-docs/tutorials/docs/2_coinbase-smart-wallet.md

```diff
@@ -0,0 +1,694 @@
+  metadata: string;
+};
+```
+
+Then add a state variable to hold the list of NFTs, and fetch them with `useReadContract`:
+
+```solidity
+export function NFTList() {
+  const account = useAccount();
+  const [nfts, setNfts] = useState<NFTData[]>([]);
+
+  const { data: nftData, refetch: refetchNftData } = useReadContract({
+    abi: contractData.abi,
+    address: contractData.address as `0x${string}`,
+    functionName: "getNFftsOwned",
+    args: [account.address],
+  });
+
+  useEffect(() => {
+    if (nftData) {
+      const newNfts = nftData as NFTData[];
+      setNfts(newNfts);
+    }
+  }, [nftData]);
+
+  return (
+    <div>
+      <h2>NFTs</h2>
+      {/* TODO */}
+    </div>
+  );
+}
+```
+
+### Interpreting the Metadata and Image
+
+Add a type and helper function to convert the base64 encoded metadata to JSON:
+
+```typescript
+type JSONMetadata = {
+  name: string;
+  description: string;
+  image: string;
+};
+
+function getJsonMetadata(nft: NFTData) {
+  const base64String = nft.metadata.split(',')[1];
+  const jsonString = atob(base64String);
+  return JSON.parse(jsonString) as JSONMetadata;
+}
+```
+
+The image is already in a format usable by `<img>` tags!
+
+### Displaying a List of NFTs
+
+Now that you can extract your metadata and image from your data, use it to build a render function for your NFTs:
+
+```typescript
+function renderNft(nft: NFTData) {
+  const metadata = getJsonMetadata(nft);
+  return (
+    <div key={nft.tokenId.toString()}>
+      <h3>{metadata.name}</h3>
+      <p>{metadata.description}</p>
+      <img src={metadata.image} alt={metadata.name} />
+    </div>
+  );
+}
+```
+
+And use it to display them:
+
+```typescript
+return (
+  <div>
+    <h2>NFTs</h2>
+    <div>{`Address: ${account.address}`}</div>
+    <div>{nfts.map((nft) => renderNft(nft))}</div>
+  </div>
+);
+```
+
+Add further styling or use a library to improve the appearance.
+
+### Testing with Smart Wallet
+
+Test with both your normal wallet in the browser, and use incognito or private mode to test with the Smart Wallet.
+
+Don't forget to mint some NFTs for the Smart Wallet address!
+
+Everything should work as expected for both.
+
+### Adding a Mint Button
+
+Import and set up functions to write to your contract and wait for the receipt:
+
+```typescript
+const { data: writeData, writeContract } = useWriteContract();
+const { data: receipt } = useWaitForTransactionReceipt({
+  hash: writeData,
+});
+```
+
+Wait for the receipt, and use it to trigger a refetch of the NFT data. Doing so will update the user's list of NFTs after they buy a new one:
+
+```typescript
+useEffect(() => {
+  if (receipt) {
+    refetchNftData();
+  }
+}, [receipt]);
+```
+
+Finally, add a button allowing the user to purchase a new NFT:
+
+```typescript
+<button
+  onClick={() =>
+    writeContract({
+      abi: contractData.abi,
+      address: contractData.address as `0x${string}`,
+      functionName: 'mintTo',
+      args: [account.address],
```

Comment:
Reminder, we are passing the address to mintTo for handling both EOA and smart contract addresses to handle msg.sender
I would double on this point since it's newer to devs